src/tests/all.rs

```diff
@@ -788,3 +775,245 @@ fn new_crate_to_body_with_tarball(new_crate: &u::NewCrate, tarball: &[u8]) -> Ve
+        self.make_request()
+    }
+
+    /// Add a user as an owner for a crate.
+    pub fn add_owner(&mut self, krate_name: &str, user: &User) {
+        let body = format!("{{\"users\":[\"{}\"]}}", user.gh_login);
+        self.request
+            .with_path(&format!("/api/v1/crates/{}/owners", krate_name))
+            .with_method(Method::Put)
+            .with_body(body.as_bytes());
+
+        let mut response = self.make_request();
+
+        #[derive(Deserialize)]
+        struct O {
+            ok: bool,
+        }
+        assert!(json::<O>(&mut response).ok);
+    }
+
+    /// As the currently logged in user, accept an invitation to become an owner of the named
+    /// crate.
+    pub fn accept_ownership_invitation(&mut self, krate_name: &str) {
+        use views::InvitationResponse;
+
+        let krate_id = {
+            let conn = self.app.diesel_database.get().unwrap();
+            Crate::by_name(krate_name)
+                .first::<Crate>(&*conn)
+                .unwrap()
+                .id
+        };
+
+        let body = json!({
+            "crate_owner_invite": {
+                "invited_by_username": "",
+                "crate_name": krate_name,
+                "crate_id": krate_id,
+                "created_at": "",
+                "accepted": true
+            }
+        });
+
+        self.request
+            .with_path(&format!("/api/v1/me/crate_owner_invitations/{}", krate_id))
+            .with_method(Method::Put)
+            .with_body(body.to_string().as_bytes());
+
+        let mut response = self.make_request();
+
+        #[derive(Deserialize)]
+        struct CrateOwnerInvitation {
+            crate_owner_invitation: InvitationResponse,
+        }
+
+        let crate_owner_invite = ::json::<CrateOwnerInvitation>(&mut response);
+        assert!(crate_owner_invite.crate_owner_invitation.accepted);
+        assert_eq!(crate_owner_invite.crate_owner_invitation.crate_id, krate_id);
+    }
+
+    /// Get the crates owned by the specified user.
+    pub fn crates_owned_by(&mut self, user: &User) -> CrateList {
+        let query = format!("user_id={}", user.id);
+        self.request
+            .with_path("/api/v1/crates")
+            .with_method(Method::Get)
+            .with_query(&query);
+
+        let mut response = self.make_request();
+
+        json::<CrateList>(&mut response)
+    }
+}
+
+lazy_static!{
+    static ref EMPTY_TARBALL_BYTES: Vec<u8> = {
+        let mut empty_tarball = vec![];
+        {
+            let mut ar = tar::Builder::new(GzEncoder::new(&mut empty_tarball, Compression::default()));
+            t!(ar.finish());
+        }
+        empty_tarball
+    };
+}
+
+/// A builder for constructing a crate for the purposes of testing publishing. If you only need
+/// a crate to exist and don't need to test behavior caused by the publish request, inserting
+/// a crate into the database directly by using CrateBuilder will be faster.
+pub struct PublishBuilder {
```

Comment:
I really like the `PublishBuilder` and `CrateBuilder` distinction here.  We currently do some strange things in our existing crate creation logic.  Porting things over to this will definitely be an improvement!